design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -399,18 +409,101 @@ are the basic configurations that the tool should support.
+
+To avoid the possibility of multiple migrations being performed at the same time, we can emulate how the
+ksqlDB command topic uses Kafka transactions to ensure that only one migration takes place at a time. 
+We'd set the producer's transactional ID based on the ksqlDB cluster's service ID (which the migrations
+tool can obtain via the `/info` endpoint of the ksqlDB server), and the flow for applying a migration would be:
+1. Begin transaction.
+1. Check `CURRENT` state is either `Migrated` or `Error` (i.e., not `Running`), by creating a Kafka consumer. Return error and abort if `Running`.
+1. Validate checksums for applied migrations
+1. Produce `Running` state to migrations stream.
+1. Commit transaction.
+1. Perform migration. 
+1. If successful, produce `Migrated`, else `Error`.
+
+The reason step 2 (checking `CURRENT` state) requires a Kafka consumer rather than simply issuing a pull query
+against the migrations table is because the pull query is not guaranteed to be up to date. (For clusters that
+have lag reporting enabled, we could specify zero lag tolerance on the pull query, but this is only possible if
+the ksqlDB cluster has lag reporting enabled, which the migrations tool cannot control.) By using a Kafka consumer
+(as does the command topic transaction protocol), we can ensure the consumer has consumed up to the end offset of 
+the topic.
+
+If we do this, I think we'll need to set `isolation.level=read_committed` on the consumer in the query that populates
+the migrations table from the migrations stream, to avoid uncommitted data from entering the pull query state stores.
+
+In the event that the migrations tool produces `Running` and crashes before completing the migration, the user will
+have to use the `abort` command on the migrations tool to transition the status of the current migration to `Error`
+in order to unblock future migrations. Note that this does not actually roll back any commands that were applied as
+part of attempting the migration; any such statements will have to be cleaned up by the user outside of the migrations tool. 
+
+#### Do not implement a transaction mechanism and accept the possibility of race conditions
+
+If we want to avoid the complexity of implementing a transaction protocol, we can accept the possibility
+of race conditions if a user attempts to perform multiple migrations at once and opt for documentation
+warning about this instead.
+
+The flow for applying a migration would be:
+1. Check `CURRENT` state is either `Migrated` or `Error` (i.e., not `Running`), by issuing a pull query. Return error and abort if `Running`.
+1. Validate checksums for applied migrations, by issuing pull queries for `status='Migrated'`.
+1. Check `CURRENT` state has not changed from above, with another pull query. Return error and abort if a change is detected.
+1. Produce `Running` state to migrations stream.
+1. Perform migration. 
+1. If successful, produce `Migrated`, else `Error`.
```

Comment:
One extra step to consider is to check (with a pull query) that a schema version has been migrated before continuing applying the next schema version. You could retry this for sometime in case there is a pull query lag. This to ensure the node where the migration process is running will be in a good state at the end.

Reply:
Yeah, the migration status flows I wrote were only for single migrations. If we're applying multiple, we can just repeat the last four steps without re-doing the initial validation.